tendermint-rs/src/account.rs

```diff
@@ -0,0 +1,100 @@
+//! Tendermint accounts
+
+use crate::error::Error;
+#[cfg(feature = "serde")]
+use serde::de::{self, Deserialize, Deserializer};
+use sha2::{Digest, Sha256};
+use signatory::ecdsa::curve::secp256k1;
+use std::{
+    fmt::{self, Display},
+    str::FromStr,
+};
+use subtle::{self, ConstantTimeEq};
+use subtle_encoding::hex;
+
+/// Size of an  account ID in bytes
+pub const ID_LENGTH: usize = 20;
+
+/// Account IDs
+#[derive(Copy, Clone, Debug, Hash)]
+pub struct Id([u8; ID_LENGTH]);
+
+impl Id {
+    /// Create a new account ID from raw bytes
+    pub fn new(bytes: [u8; ID_LENGTH]) -> Id {
+        Id(bytes)
+    }
+
+    /// Borrow the account ID as a byte slice
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0[..]
+    }
+}
+
+impl AsRef<[u8]> for Id {
+    fn as_ref(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
+
+impl ConstantTimeEq for Id {
+    #[inline]
+    fn ct_eq(&self, other: &Id) -> subtle::Choice {
+        self.as_bytes().ct_eq(other.as_bytes())
+    }
+}
+
+impl Display for Id {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for byte in &self.0 {
+            write!(f, "{:02X}", byte)?;
+        }
+        Ok(())
+    }
+}
+
+impl From<secp256k1::PublicKey> for Id {
```

Comment:
This implementation is copy/paste from `node::Id`, but I assume account IDs/addresses work the same way? Are there test vectors I can use?

Should this be called `account::Address` instead?

Reply:
> but I assume account IDs/addresses work the same way? 

Yes, the only difference is the curve (for now). secp vs ed2219

> Are there test vectors I can use?

Does this work? 
https://github.com/tendermint/tendermint/blob/5b8888b01b7542cd6dd856e7a322f3d4e5ccba16/crypto/secp256k1/secp256k1_test.go#L24-L30

> Should this be called account::Address instead?

Good question. Tendermint doesn't have these semantics (Address is the truncated hash of whatever pubkey given, ID is only for nodes). Gaia has account addresses, but it is the bechified version: https://github.com/cosmos/cosmos-sdk/blob/9cdd1d3e9ebf5fdb72203acbaf6231c5dcde82b9/types/address.go#L73

Reply:
Will add that test vector to this PR, thanks!

Reply:
Looking at the test vectors in the address code you linked, I can't tell what format the `addr` is (it's mixed case so possibly Base64?)

The format I'm looking at (i.e. the `validator_info` object's `address` field in the `/status` RPC endpoint) appears to be hex, and looks like:

```
C73833E9BD86D34EDAD4AFD571FB5D0926294CD5
```

Reply:
> I can't tell what format that is. It's mixed case so possibly Base64?

Looks like it is base58 (I should have provided the following lines for context too):
https://github.com/tendermint/tendermint/blob/5b8888b01b7542cd6dd856e7a322f3d4e5ccba16/crypto/secp256k1/secp256k1_test.go#L36